core/server/common/src/main/java/alluxio/master/BackupManager.java

```diff
@@ -63,18 +74,139 @@ public BackupManager(MasterRegistry registry) {
    * @param os the stream to write to
    */
   public void backup(OutputStream os) throws IOException {
-    int count = 0;
+    // Create gZIP compressed stream as back-up stream.
     GzipCompressorOutputStream zipStream = new GzipCompressorOutputStream(os);
-    for (Master master : mRegistry.getServers()) {
-      Iterator<JournalEntry> it = master.getJournalEntryIterator();
-      while (it.hasNext()) {
-        it.next().toBuilder().clearSequenceNumber().build().writeDelimitedTo(zipStream);
-        count++;
+
+    // Executor for writing backup content.
+    ExecutorService backupWriterExecutor =
+        Executors.newSingleThreadExecutor(ThreadFactoryUtils.build("backup-writer-%d", true));
+    // Executor for reading master state.
+    ExecutorService backupMasterReaderExecutor = Executors
+        .newSingleThreadExecutor(ThreadFactoryUtils.build("backup-master-reader-%d", true));
+
+    // Entry queue will be used as a buffer and synchronization between readers and writer.
+    // Use of {@link LinkedBlockingQueue} is preferred becaue of {@code #drainTo()} method,
+    // using which all existing entries can be drained while allowing writes.
+    // Processing/draining one-by-one using {@link ConcurrentLinkedQueue} proved to be
+    // inefficient compared to draining with dedicated method.
+    LinkedBlockingQueue<JournalEntry> journalEntryQueue = new LinkedBlockingQueue<>(
+        ServerConfiguration.getInt(PropertyKey.MASTER_BACKUP_ENTRY_BUFFER_COUNT));
+
+    // Shows how many entries have been written.
+    AtomicLong writtenEntryCount = new AtomicLong(0);
+
+    // Futures for reader/writer tasks.
+    SettableFuture<Void> readerTaskFuture = SettableFuture.create();
+    SettableFuture<Void> writerTaskFuture = SettableFuture.create();
+
+    // Submit master reader task.
+    backupMasterReaderExecutor.submit(() -> {
+      try {
+        for (Master master : mRegistry.getServers()) {
+          Iterator<JournalEntry> it = master.getJournalEntryIterator();
+          if (it != null) {
+            while (!writerTaskFuture.isDone() && it.hasNext()) {
+              JournalEntry je = it.next();
```

Comment:
Any attempt to guarantee a deterministic answer for hasNext() will need a synchronization with buffering process. Since hasNext() is called as the same amount as next(), any synchronization will cause a contention. I don't see a way without badly hurting the concurrency of buffering.

Reply:
I'm thinking of pushing the "it might be null" handling code into the iterator instead of here. The synchronization shouldn't need to change

Reply:
I think it could done by an additional buffering of next entry and moving blocking part from next() to hasNext().